cpp/src/binary/jit/core/binop.cpp

```diff
@@ -0,0 +1,229 @@
+        GDF_REQUIRE((out->size == rhs->size), GDF_COLUMN_SIZE_MISMATCH)
+
+        // Check for null data pointer
+        GDF_REQUIRE((out->data != nullptr) && (rhs->data != nullptr), GDF_DATASET_EMPTY)
+
+        // Check for datatype
+        GDF_REQUIRE((out->dtype > GDF_invalid) && (lhs->dtype > GDF_invalid) && (rhs->dtype > GDF_invalid), GDF_UNSUPPORTED_DTYPE)
+        GDF_REQUIRE((out->dtype < N_GDF_TYPES) && (lhs->dtype < N_GDF_TYPES) && (rhs->dtype < N_GDF_TYPES), GDF_UNSUPPORTED_DTYPE)
+
+        scalar_col_valid_mask_and(out->null_count, out->valid, rhs->valid, lhs->is_valid, rhs->size);
+
+        Launcher::launch().kernel("kernel_v_s")
+                          .instantiate(ope, Operator::Type::Reverse, out, rhs, lhs)
+                          .launch(out, rhs, lhs);
+
+        return GDF_SUCCESS;
+    }
+
+    gdf_error binary_operation(gdf_column* out, gdf_column* lhs, gdf_scalar* rhs, gdf_binary_operator ope) {
+
+        // Check for null pointers in input
+        GDF_REQUIRE((out != nullptr) && (lhs != nullptr) && (rhs != nullptr), GDF_DATASET_EMPTY)
+
+        // Check for 0 sized data
+        GDF_REQUIRE((out->size != 0) && (lhs->size != 0), GDF_SUCCESS)
+        GDF_REQUIRE((out->size == lhs->size), GDF_COLUMN_SIZE_MISMATCH)
+
+        // Check for null data pointer
+        GDF_REQUIRE((out->data != nullptr) && (lhs->data != nullptr), GDF_DATASET_EMPTY)
+
+        // Check for datatype
+        GDF_REQUIRE((out->dtype > GDF_invalid) && (lhs->dtype > GDF_invalid) && (rhs->dtype > GDF_invalid), GDF_UNSUPPORTED_DTYPE)
+        GDF_REQUIRE((out->dtype < N_GDF_TYPES) && (lhs->dtype < N_GDF_TYPES) && (rhs->dtype < N_GDF_TYPES), GDF_UNSUPPORTED_DTYPE)
+
+        scalar_col_valid_mask_and(out->null_count, out->valid, lhs->valid, rhs->is_valid, lhs->size);
+
+        Launcher::launch().kernel("kernel_v_s")
+                          .instantiate(ope, Operator::Type::Direct, out, lhs, rhs)
+                          .launch(out, lhs, rhs);
+
+        return GDF_SUCCESS;
+    }
+
+    gdf_error binary_operation(gdf_column* out, gdf_column* lhs, gdf_column* rhs, gdf_binary_operator ope) {
+
+        // Check for null pointers in input
+        GDF_REQUIRE((out != nullptr) && (lhs != nullptr) && (rhs != nullptr), GDF_DATASET_EMPTY)
+
+        // Check for 0 sized data
+        GDF_REQUIRE((out->size != 0) && (lhs->size != 0) && (rhs->size != 0), GDF_SUCCESS)
+        GDF_REQUIRE((out->size == lhs->size) && (lhs->size == rhs->size), GDF_COLUMN_SIZE_MISMATCH)
+
+        // Check for null data pointer
+        GDF_REQUIRE((out->data != nullptr) && (lhs->data != nullptr) && (rhs->data != nullptr), GDF_DATASET_EMPTY)
+
+        // Check for datatype
+        GDF_REQUIRE((out->dtype > GDF_invalid) && (lhs->dtype > GDF_invalid) && (rhs->dtype > GDF_invalid), GDF_UNSUPPORTED_DTYPE)
+        GDF_REQUIRE((out->dtype < N_GDF_TYPES) && (lhs->dtype < N_GDF_TYPES) && (rhs->dtype < N_GDF_TYPES), GDF_UNSUPPORTED_DTYPE)
```

Comment:
There seems to be a lot of redundancy with the scalar version...
Also, consider placing the requirement checks in a separate function.

Reply:
Out of 6 conditions, 3 are same and 3 are different. And for those that are the same, they don't have a logical grouping.